src/core/dataset_container.jl

```diff
@@ -0,0 +1,179 @@
+    return DatasetContainer(
+        Dict{ConstraintKey, T}(),
+        Dict{AuxVarKey, T}(),
+        Dict{VariableKey, T}(),
+        Dict{ParameterKey, T}(),
+        Dict{ExpressionKey, T}(),
+    )
+end
+
+function Base.empty!(container::DatasetContainer)
+    for field in fieldnames(DatasetContainer)
+        field_dict = getfield(container, field)
+        for val in values(field_dict)
+            empty!(val)
+        end
+    end
+    @debug "Emptied the store" _group = LOG_GROUP_SIMULATION_STORE
+    return
+end
+
+function get_duals_values(container::DatasetContainer)
+    return container.duals
+end
+
+function get_aux_variables_values(container::DatasetContainer)
+    return container.aux_variables
+end
+
+function get_variables_values(container::DatasetContainer)
+    return container.variables
+end
+
+function get_parameters_values(container::DatasetContainer)
+    return container.parameters
+end
+
+function get_expression_values(container::DatasetContainer)
+    return container.expressions
+end
+
+function get_dataset_keys(container::DatasetContainer)
+    return Iterators.flatten(
+        keys(getfield(container, f)) for f in fieldnames(DatasetContainer)
+    )
+end
+
+function get_dataset(container::DatasetContainer, key::OptimizationContainerKey)
+    datasets = getfield(container, get_store_container_type(key))
+    return datasets[key]
+end
+
+function set_dataset!(
+    container::DatasetContainer{T},
+    key::OptimizationContainerKey,
+    val::T,
+) where {T <: AbstractDataset}
+    datasets = getfield(container, get_store_container_type(key))
+    datasets[key] = val
+    return
+end
+
+function has_dataset(container::DatasetContainer, key::OptimizationContainerKey)
+    datasets = getfield(container, get_store_container_type(key))
+    return haskey(datasets, key)
+end
+
+function get_dataset(
+    container::DatasetContainer,
+    ::T,
+    ::Type{U},
+) where {T <: ConstraintType, U <: Union{PSY.Component, PSY.System}}
+    return get_dataset(container, ConstraintKey(T, U))
+end
+
+function get_dataset(
+    container::DatasetContainer,
+    ::T,
+    ::Type{U},
+) where {T <: VariableType, U <: Union{PSY.Component, PSY.System}}
+    return get_dataset(container, VariableKey(T, U))
+end
+
+function get_dataset(
+    container::DatasetContainer,
+    ::T,
+    ::Type{U},
+) where {T <: AuxVariableType, U <: Union{PSY.Component, PSY.System}}
+    return get_dataset(container, AuxVarKey(T, U))
+end
+
+function get_dataset(
+    container::DatasetContainer,
+    ::T,
+    ::Type{U},
+) where {T <: ParameterType, U <: Union{PSY.Component, PSY.System}}
+    return get_dataset(container, ParameterKey(T, U))
+end
+
+function get_dataset(
+    container::DatasetContainer,
+    ::T,
+    ::Type{U},
+) where {T <: ExpressionType, U <: Union{PSY.Component, PSY.System}}
+    return get_dataset(container, ExpressionKey(T, U))
+end
+
+function get_dataset_values(container::DatasetContainer, key::OptimizationContainerKey)
```

Comment:
Same comment about these functions and different return types.

Reply:
This function is more problematic, I don't have a good alternative in this case unless we always return an array.

Reply:
We could always return either a dataframe or an array.

Reply:
Still pending what the requirements are until we have the HDF data set implemented.